DiscordChatExporter.Core/Exporting/Filtering/Parsing/FilterParser.cs

```diff
@@ -0,0 +1,76 @@
+﻿using Superpower;
+using Superpower.Model;
+using Superpower.Parsers;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text;
+
+namespace DiscordChatExporter.Core.Exporting.Filtering.Parsing
+{
+    public static class FilterParser
+    {
+        public static TokenListParser<FilterToken, string> AnyString { get; } =
+            Token.EqualTo(FilterToken.QuotedString).Apply(FilterTextParsers.QuotedString)
+            .Or(Token.EqualTo(FilterToken.UnquotedString).Apply(FilterTextParsers.UnquotedString));
+
+        public static TokenListParser<FilterToken, MessageFilter> AnyFilter { get; } =
+            from minus in Token.EqualTo(FilterToken.Minus).Optional()
+            from content in KeyValuePair.Or(TextFilter.Try()).Or(GroupedFilter)
+            select minus.HasValue ? new NegatedMessageFilter(content) : content;
+
+        public static TokenListParser<FilterToken, MessageFilter> KeyValuePair { get; } =
+            from key in AnyString.Try()
+            from colon in Token.EqualTo(FilterToken.Colon).Try()
+            from value in AnyString
+            select MessageFilter.CreateFilter(key, value);
+
+        public static TokenListParser<FilterToken, MessageFilter> TextFilter { get; } =
+            from value in AnyString
+            select MessageFilter.CreateFilter(value);
+
+        public static TokenListParser<FilterToken, MessageFilter> MultipleFilter { get; } =
+            from first in AnyFilter
+            from rest in Token.EqualTo(FilterToken.VBar)
+                .IgnoreThen(AnyFilter)
+                .Select(value => (Value: value, Unioned: true)) //anonymous struct keeps track of whether a vertical bar was used
+                .Or(AnyFilter.Select(value => (Value: value, Unioned: false)))
+                .Many()
```

Comment:
I would do it a bit differently:

1. Have two separate parsers
2. One parser for || expression -> `OrBinaryExpressionFilter`
3. One parser for && expression -> `AndBinaryExpressionFilter`
4. Then have a higher-kind parser for both -> `OrBinaryExpressionFilter.Or(AndBinaryExpressionFilter).Many()` (or something like that)

Does it make sense?